pkg/apiserver/rest/dualwriter_mode2.go

```diff
@@ -0,0 +1,182 @@
+package rest
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/klog"
+)
+
+type DualWriterMode2 struct {
+	DualWriter
+}
+
+// NewDualWriterMode2 returns a new DualWriter in mode 2.
+func NewDualWriterMode2(legacy LegacyStorage, storage Storage) *DualWriterMode2 {
+	return &DualWriterMode2{*newDualWriter(legacy, storage)}
+}
+
+// Get overrides the default behavior of the Storage and retrieves an object from
+// LegacyStorage or Storage depending on the DualWriter mode.
+func (d *DualWriterMode2) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	legacy, ok := d.legacy.(rest.Getter)
+	if !ok {
+		return nil, fmt.Errorf("legacy storage rest.Getter is missing")
+	}
+
+	s, err := d.Storage.Get(ctx, name, &metav1.GetOptions{})
+	if err == nil {
+		return s, err
+	}
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	return legacy.Get(ctx, name, &metav1.GetOptions{})
```

Comment:
We said Mode 2 means reading from legacy storage. Two reasons I can think of for this:
* the object might not exist in k8s storage
* we aren't confident about the k8s storage object having the correct values set

This approach addresses the first issue. If the second point is a concern, we could focus on testing.